src/gui/render/graphics_shaders/program_handler.cpp

```diff
@@ -0,0 +1,340 @@
+    if (info_log_length > 0) {
+        std::string shader_error_message(info_log_length + 1, '\0');
+        glGetShaderInfoLog(
+            shader_ID_, info_log_length, nullptr, shader_error_message.data()
+        );
+
+        LOG_ERROR(
+            logging::opengl_logger, "{} shader error: {}", shader_type_string,
+            shader_error_message
+        );
+        status_ = ShaderStatus::COMPILATION_ERROR;
+        return;
+    }
+
+    status_ = ShaderStatus::OK;
+}
+
+void
+Program::reload() {
+    found_uniforms_.clear();
+    uniforms_.clear();
+
+    if (vertex_shader_.get_status() != ShaderStatus::OK) {
+        // Not ok reload
+        vertex_shader_.reload();
+        // second fail
+        if (vertex_shader_.get_status() != ShaderStatus::OK) {
+            LOG_ERROR(logging::opengl_logger, "Some error occurred.");
+            status_ = ProgramStatus::INVALID_SHADER;
+            return;
+        }
+    }
+    if (fragment_shader_.get_status() != ShaderStatus::OK) {
+        // Not ok reload
+        fragment_shader_.reload();
+        // second fail
+        if (fragment_shader_.get_status() != ShaderStatus::OK) {
+            LOG_ERROR(logging::opengl_logger, "Some error occurred.");
+            status_ = ProgramStatus::INVALID_SHADER;
+            return;
+        }
+    }
+
+    found_uniforms_.insert(
+        vertex_shader_.uniform_begin(), vertex_shader_.uniform_end()
+    );
+    found_uniforms_.insert(
+        fragment_shader_.uniform_begin(), fragment_shader_.uniform_end()
+    );
+
+    GLint vertex_shader_id = vertex_shader_.get_shader_ID();
+    GLint fragment_shader_id = fragment_shader_.get_shader_ID();
+
+    GLint Result = GL_FALSE;
+    int info_log_length;
+
+    // Link the program
+    // LOG_BACKTRACE(logging::opengl_logger, "Linking shader program");
+
+    program_ID_ = glCreateProgram();
+    glAttachShader(program_ID_, vertex_shader_id);
+    glAttachShader(program_ID_, fragment_shader_id);
+    glLinkProgram(program_ID_);
+
+    // Check the program
+    glGetProgramiv(program_ID_, GL_LINK_STATUS, &Result);
+    glGetProgramiv(program_ID_, GL_INFO_LOG_LENGTH, &info_log_length);
+    if (info_log_length > 0) {
+        std::string program_error_message(info_log_length + 1, '\0');
+        glGetProgramInfoLog(
+            program_ID_, info_log_length, nullptr, program_error_message.data()
+        );
+
+        LOG_ERROR(
+            logging::opengl_logger, "Shader program linking error: {}",
+            program_error_message
+        );
+        status_ = ProgramStatus::LINKING_FAILED;
+        goto detach_shader;
+    }
+
+    LOG_INFO(
+        logging::opengl_logger, "Shader compiled successfully with program ID {}",
+        program_ID_
+    );
+
+    status_ = ProgramStatus::OK;
+    attach_uniforms();
+
+detach_shader:
+
+    glDetachShader(program_ID_, vertex_shader_id);
+    glDetachShader(program_ID_, fragment_shader_id);
```

Comment:
why do you detach the shader on success?

Reply:
This may be only on your impl of openGL

Reply:
https://gamedev.stackexchange.com/questions/47910/after-a-succesful-gllinkprogram-should-i-delete-detach-my-shaders

Reply:
Well then you should also `glDeleteShader` on success too. That's what that SO post said.

Reply:
Multiple programs use the same shader. That's why they are not deleted.